pkg/apis/core/annotation_key_constants.go

```diff
@@ -122,8 +122,21 @@ const (
 	// This annotation is beta-level and is only honored when PodDeletionCost feature is enabled.
 	PodDeletionCost = "controller.kubernetes.io/pod-deletion-cost"
 
-	// AnnotationTopologyAwareHints can be used to enable or disable Topology
-	// Aware Hints for a Service. This may be set to "Auto" or "Disabled". Any
-	// other value is treated as "Disabled".
-	AnnotationTopologyAwareHints = "service.kubernetes.io/topology-aware-hints"
+	// DeprecatedAnnotationTopologyAwareHints can be used to enable or disable
+	// Topology Aware Hints for a Service. This may be set to "Auto" or
+	// "Disabled". Any other value is treated as "Disabled". This annotation has
+	// been deprecated in favor of the `service.kubernetes.io/topology-mode`
+	// annotation. When both `service.kubernetes.io/topology-aware-hints` and
+	// `service.kubernetes.io/topology-mode` annotations are set, the value of
+	// `service.kubernetes.io/topology-mode` has precedence.
+	DeprecatedAnnotationTopologyAwareHints = "service.kubernetes.io/topology-aware-hints"
+
+	// AnnotationTopologyMode can be used to enable or disable Topology Aware
+	// Routing for a Service. Well known values are "Auto" and "Disabled".
+	// Implementations may choose to develop new topology approaches, exposing
+	// them with domain-prefixed values. For example, "example.com/lowest-rtt"
+	// could be a valid implementation-specific value for this annotation. These
+	// heuristics will often populate topology hints on EndpointSlices, but that
+	// is not a requirement.
```

Comment:
EndpointSlice hints are a typed field so implementation-specific topology modes would not be able to use them, right? Does that put implementation-specific topology modes at a disadvantage relative to official ones?

For that matter, why is this an annotation but the EndpointSlice Hints are a field?

Reply:
>EndpointSlice hints are a typed field so implementation-specific topology modes would not be able to use them, right? 

I think this is all working towards a place where people that want to experiment with new hints heuristics can bring their own fork of EPS controller. That would be made easier by https://github.com/kubernetes/enhancements/pull/3686 and could be another reason to invest in a KEP for something like https://github.com/kubernetes/kubernetes/pull/87488. The conclusion I've personally come to is that it's going to be very hard to find a heuristic that works for everyone. Providing a way for the community to develop others seems like it may be our best path forward. In the future, if there are clearly effective and widely used community heuristics, we may want to merge them into k/k.

>For that matter, why is this an annotation but the EndpointSlice Hints are a field?

Probably the most concrete reason is that it's not possible to attach annotations to a specific endpoint in the list. We're also inching closer to a place where the dataplane doesn't need to care why a hint is present, just that it is. (Prior to this PR the annotation needed to be set to "Auto", now it just can't be set to "Disabled").

Open to other ideas here, but at a high level I think it's important that we provide a way for community heuristics to be developed and experimented with because I don't think we can make the correct routing decisions for everyone with a single algorithm and/or the limited time we have to invest in this feature.

Reply:
cc @s1061123 and @akhilles since this relates to ideas you've worked on

Reply:
> Does that put implementation-specific topology modes at a disadvantage relative to official ones?

I think that, if there was a need and it was reasonable, we could consider additional hints.  Unfortunately, our API and our implementation are entangled and we don't have an awesome way to extend existing resources.

In a perfect world, hints would be an extension of the base type, but there's just no way to express that.  We could put all hints in a parallel type, but now we have wonky races where the slice is updated but the hints are not, and we 2x the API ops.  We could render hints as a per-slice annotation.  That's probably the "fairest" approach, but ick.

I expect that other implementations won't use EPSlice anyway, so it's moot

> why is this an annotation but the EndpointSlice Hints are a field

This is fair - `topologyMode` could be a field.  As much churn as this has undergone, that would require resetting to alpha.  I think I'd just as soon run with annotation and if this pattern sticks, we can eat the tech-debt of converting to a field.

Agree or disagree?

Reply:
> I think I'd just as soon run with annotation and if this pattern sticks, we can eat the tech-debt of converting to a field.
> 
> Agree or disagree?

agree

Reply:
@robscott thank you for the mention to me. My interest is delegate endpointSlice management for specific service, by other controller, that is not to create any endpointslices by EPS controller in Kubernetes controller manager. In this case, this `topology-mode` annotation only controls the topology, so I believe that endpointSlice is created. So it is slightly related but not so related to me.